data-structures/data.md

```diff
@@ -0,0 +1,30 @@
+`Flexible Byte Layout` is an advanced layout for representing binary data.
+
+It is flexible enough to support very small and very large (multi-block) binary data.
+
+```sh
+type Lengths [Int]
+type ByteUnionList [&FlexibleByteLayout]
```

Comment:
this `&` is a bit annoying, maybe we should be introducing `link` into a kinded union somewhere to make this linking optional, cause this can't be properly fully inlined where `parts` is used

Reply:
I wonder if we could extend `FlexibleByteLayout` to be recursive:

```

type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList map
  | FlexibleByteLayout link
} representation kinded
```

I _think_ that's allowed and reasonable since this is only really a type hint that can't be strongly enforced at the individual block level prior to loading the link

Reply:
^ if we did that, then we could drop the `&` here and we get fully-inline as an option, I think?

Reply:
```
type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList map
  | &FlexibleByteLayout link
} representation kinded
```

I think might be the correct way of doing it